fix: clone session data on .set()

kaytwo opened the pull request:
## Changes

* Fixes #12881 by saving a clone of the set()'d object at time of set()
* Refactor unflatten and stringify so that the stringify check works for URL objects


I chose to use `parse(stringify(value))` instead of anything like `structuredClone` or trying to get fancy with the store to ensure correctness, the extra parse should not be a significant performance issue unless people start storing pathologically large session data.

## Testing

Added a new test in `sessions.test.js` that fails on main and succeeds on this branch.

That test could probably use a helper function to maintain a "cookie jar" for session, but I don't quite know the cleanest way to do that.


## Docs

No update necessary, just a bugfix. However if the powers that be decide #12881 is `wontfix`, the docs definitely need to be updated to remind users to clone their objects before calling set, or be careful with updating them after doing so.

<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
<!-- /cc @withastro/maintainers-docs for feedback! -->

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

ematipico commented on packages/astro/src/core/session.ts:
```diff
@@ -138,9 +152,10 @@ export class AstroSession<TDriver extends SessionDriverName = any> {
 		}
+		let cloned: T;
 		try {
 			// Attempt to serialize the value so we can throw an error early if needed
```
The comment should be updated to reflect the new logic

ascorbic replied:
Agreed. Can you include the explanation of why we're cloning it too.

ascorbic commented on packages/astro/test/sessions.test.js:
```diff
@@ -1,13 +1,17 @@
 
 	before(async () => {
+		// seems to be a race condition in unstorage when the session directory doesn't exist
+		await fs.promises.mkdir('.astro/session', { recursive: true });
```
We should probably be creating this in production too in that case. It could go here: 

https://github.com/kaytwo/astro/blob/6e7afe876fa8ee07457dd60e145de70aa3f9f5c3/packages/astro/src/core/session.ts#L396

ascorbic commented:
Thanks for this. The issue with the missing `.astro/sessions` dir could be a real bug, so it would be better to fix it than work around it in the test.

ascorbic commented:
Ah, my mistake. We can't use fs there. I think it's ok to revert to your previous version.

kaytwo commented:
reverted. I added a different "ensure the directory exists" check that is only a little bit of a hack. also added "/.astro" to .gitignore as both manually and turbo running the tests creates that directory in various places, and I didn't see anywhere else (e.g. being used as part of a test fixture) so that should be safe.

ascorbic commented on packages/astro/src/core/session.ts:
```diff
@@ -455,6 +449,13 @@ export class AstroSession<TDriver extends SessionDriverName = any> {
 			this.#storage = createStorage({
 				driver: driver(this.#config.options),
 			});
+			if (this.#config.driver === 'fs-lite') {
+				// seems to be a race condition in unstorage when the session directory doesn't exist
+				// this ensures that the directory is created before we try to use it for real sessions
+				const dummyId = crypto.randomUUID();
+				await this.#storage.setItem(dummyId, 'init');
+				await this.#storage.removeItem(dummyId);
+			}
```
OK, I think I understand this now. SAo is the problem that if you try to read a session when the ddirectory hasn't been created then it has an error? I'm not happy with this workaround because we're otherwise not using the storage unless the user actually reads or writes a session. This is soemthign that should be fixed upstream: unstorage should just return undefined for gets if the directory doesn't exist, the same as if the file doesn't exist. 

Have you encountered this issue outside of tests? If not, I think the best bet is to add a workaround inside tests, and report the issue on the unstorage repo.

kaytwo replied:
Glad you pushed back on this one - found the underlying bug. `PERSIST_SYMBOL` snuck through TypeScript's type inference and it didn't yell about it not being awaited.

ascorbic commented on .changeset/tough-cars-yawn.md:
```diff
@@ -0,0 +1,5 @@
+'astro': patch
+---
+
+Fixes a bug where session data could be corrupted if it is changed after calling .set(); also fixes a bug where responses can be returned before session data is saved
```
Can you split this into two chanegsets. You can run `pnpm changeset` again to add another one.

ascorbic reviewed (approved):
Great work. Thanks for tracking down all the issues. Just a single request with the changeset and we can merge this.

ascorbic reviewed (approved):
Great! Thanks a lot for fixing these two.
